application/controllers/admin/tokens.php

```diff
@@ -2148,7 +2148,42 @@ public function import($iSurveyId)
                             }
                         }
 
-                        if (!$bDuplicateFound && !$bInvalidEmail && !$bInvalidToken) {
+                        // Dispatch beforeTokenImport event
+                        $params = array(
+                            'csvcharset' => $sUploadCharset,
+                            'filterduplicatetoken' => $bFilterDuplicateToken,
+                            'filterblankemail' => $bFilterBlankEmail,
+                            'allowinvalidemail' => $bAllowInvalidEmail,
+                            'filterduplicatefields' => $aFilterDuplicateFields,
+                            'separator' => $sSeparator,
+                            'showwarningtoken' => Yii::app()->request->getPost('showwarningtoken'),
+                        );
+                        $event = new PluginEvent('beforeTokenImport');
+                        $event->set('importType', 'CSV');
+                        $event->set('surveyId', $iSurveyId);
+                        $event->set('params', $params);
+                        $event->set('recordCount', $iRecordCount);
+                        $event->set('token', $aWriteArray);
+                        App()->getPluginManager()->dispatchEvent($event);
+                        $bPluginReportedError = !is_null($event->get('importValid')) && !$event->get('importValid');
```

Comment:
Why not add $event->set('importValid', true) bedore dispatching event. Then after event we can just test `if(!$event->get('importValid'))` ?